Add user metadata

maeldebon opened the pull request:
🔗 **Related PRs:**
- [substra-tests](https://github.com/SubstraFoundation/substra-tests/pull/90)
- [substra](https://github.com/SubstraFoundation/substra/pull/143)
- [substra-backend](https://github.com/SubstraFoundation/substra-backend/pull/186)
- [substra-frontend (coming soon)](#)

The goal of these PRs is to add a `metadata` field in every asset in order to be able to store information related to this asset. It could be, for instance, a creation_ts, an epoch, a project_name, ...

Commit:
Add a metadata field

maeldebon commented on chaincode/algo_aggregate_test.go:
```diff
@@ -16,6 +16,7 @@ package main
 
 import (
 	"encoding/json"
+	"fmt"
```
Oops, shouldn't be here, I'll remove this in the next commit

Commit:
Update tests

thibaultrobert commented on chaincode/input.go:
```diff
@@ -31,42 +31,46 @@ var (
+	MetricsStorageAddress     string           	`validate:"required,url" json:"metricsStorageAddress"`
+	TestDataset               inputDataset     	`validate:"omitempty" json:"testDataset"`
+	Permissions               inputPermissions 	`validate:"required" json:"permissions"`
+	Metadata				  map[string]string	`validate:"omitempty" json:"metadata"`
```
It's strange that your editor didn't take care of it but usually the different section of a struct validation are aligned.

thibaultrobert replied:
Got the problem: you've used tabs between names, types, and tags. It should have been spaces.

thibaultrobert replied:
I formatted all go file, see the diff: https://gist.github.com/thibaultrobert/1e142684add2418739b6793c0a3103b6
You can apply it or try to tweak the formatting tools of your editor/IDE.

maeldebon replied:
Yes, thank you! It should be better like this

Commit:
Fix spacing

Replace tabs by spaces // Fix order

Commit:
Add maximum length in input

Commit:
Add metadata to compute plans

Add a metadata field to the compute plans // Rework validation conditions for metadata // Fix tests

Commit:
Fix tests

Commit:
Fix spacing

Commit:
Remove flatten metadata

Commit:
Add initMapOutput to avoid none values

Commit:
Remove metadata for comput_plan

samlesu commented on chaincode/input.go:
```diff
@@ -31,42 +31,46 @@ var (
-	DataSampleKeys []string `validate:"omitempty,dive,len=64,hexadecimal" json:"dataSampleKeys"`
+	DataManagerKey string             `validate:"omitempty,len=64,hexadecimal" json:"dataManagerKey"`
+	DataSampleKeys []string           `validate:"omitempty,dive,len=64,hexadecimal" json:"dataSampleKeys"`
+	Metadata       map[string]string  `validate:"omitempty,lte=100,dive,keys,lte=50,endkeys,lte=100" json:"metadata"`
```
As far as I know, inputDataset represents an internal object and not a dataset asset. In the chaincode, the dataset asset is identified as `dataManager`.
@thibaultrobert could you confirm?

thibaultrobert replied:
Yes you right it's only use as a part of an `inputObjective`. It represent the combination of one `dataManager` and several `dataSample`.
Adding a Metadata field seems a bit overkill.

thibaultrobert committed:
Fix tests

thibaultrobert reviewed (changes requested):
I think `initMapOutput` should be used to make sure the `Metadata` fields is set in the output only, not in the ledger itself.

Commit:
Move initMapOutput in output

thibaultrobert commented on EXAMPLES.md:
```diff
@@ -292,6 +298,7 @@ peer chaincode query -n mycc -c '{"Args":["queryObjectives"]}' -C myc
     "bb1bb7c31f62244c0f3a761cc168804227115793d01c270021fe3f7935482dcc",
     "bb2bb7c31f62244c0f3a761cc168804227115793d01c270021fe3f7935482dcc"
    ],
+   "metadata": null,
```
You've still have some null metadata in output

thibaultrobert commented on EXAMPLES.md:
```diff
@@ -383,12 +392,14 @@ peer chaincode invoke -n mycc -c '{"Args":["queryFilter","{\"indexName\":\"train
     "aa1bb7c31f62244c0f3a761cc168804227115793d01c270021fe3f7935482dcc",
     "aa2bb7c31f62244c0f3a761cc168804227115793d01c270021fe3f7935482dcc"
    ],
+   "metadata": null,
```
Here too.

thibaultrobert commented on EXAMPLES.md:
```diff
@@ -481,12 +494,14 @@ peer chaincode invoke -n mycc -c '{"Args":["logSuccessTrain","{\"key\":\"ebbf6cd
    "aa1bb7c31f62244c0f3a761cc168804227115793d01c270021fe3f7935482dcc",
    "aa2bb7c31f62244c0f3a761cc168804227115793d01c270021fe3f7935482dcc"
   ],
+  "metadata": null,
```
Here too.

thibaultrobert commented on EXAMPLES.md:
```diff
@@ -430,12 +441,14 @@ peer chaincode invoke -n mycc -c '{"Args":["logStartTrain","{\"key\":\"ebbf6cdde
    "aa1bb7c31f62244c0f3a761cc168804227115793d01c270021fe3f7935482dcc",
    "aa2bb7c31f62244c0f3a761cc168804227115793d01c270021fe3f7935482dcc"
   ],
+  "metadata": null,
```
Here too.

thibaultrobert commented on EXAMPLES.md:
```diff
@@ -530,12 +545,14 @@ peer chaincode invoke -n mycc -c '{"Args":["queryTraintuple","{\"key\":\"ebbf6cd
    "aa1bb7c31f62244c0f3a761cc168804227115793d01c270021fe3f7935482dcc",
    "aa2bb7c31f62244c0f3a761cc168804227115793d01c270021fe3f7935482dcc"
   ],
+  "metadata": null,
```
Here too.

thibaultrobert commented on EXAMPLES.md:
```diff
@@ -1060,12 +1090,14 @@ peer chaincode query -n mycc -c '{"Args":["queryModelDetails","{\"key\":\"ebbf6c
     "aa1bb7c31f62244c0f3a761cc168804227115793d01c270021fe3f7935482dcc",
     "aa2bb7c31f62244c0f3a761cc168804227115793d01c270021fe3f7935482dcc"
    ],
+   "metadata": null,
```
Here too.

thibaultrobert commented on EXAMPLES.md:
```diff
@@ -1104,12 +1136,14 @@ peer chaincode query -n mycc -c '{"Args":["queryModels"]}' -C myc
      "aa1bb7c31f62244c0f3a761cc168804227115793d01c270021fe3f7935482dcc",
      "aa2bb7c31f62244c0f3a761cc168804227115793d01c270021fe3f7935482dcc"
     ],
+    "metadata": null,
```
Here too.

thibaultrobert commented on EXAMPLES.md:
```diff
@@ -1139,6 +1173,7 @@ peer chaincode query -n mycc -c '{"Args":["queryModels"]}' -C myc
      "aa1bb7c31f62244c0f3a761cc168804227115793d01c270021fe3f7935482dcc",
      "aa2bb7c31f62244c0f3a761cc168804227115793d01c270021fe3f7935482dcc"
     ],
+    "metadata": null,
```
Here too.

thibaultrobert commented on EXAMPLES.md:
```diff
@@ -1497,6 +1536,7 @@ peer chaincode invoke -n mycc -c '{"Args":["queryObjectiveLeaderboard","{\"objec
     "bb1bb7c31f62244c0f3a761cc168804227115793d01c270021fe3f7935482dcc",
     "bb2bb7c31f62244c0f3a761cc168804227115793d01c270021fe3f7935482dcc"
    ],
+   "metadata": null,
```
And here.

thibaultrobert reviewed (changes requested):
Sorry as stated in my comments you've still have some metadata that are at `null` in the output.  It seems due to embedded structures so it might not be as critical as the others but it's still incoherent.

thibaultrobert committed:
Initialize missed Metadata

thibaultrobert committed:
Fix tests

Commit:
Little fix
